README.md
# Modcolle

[![Build Status](https://travis-ci.org/makemek/Modcolle.svg?branch=dev)](https://travis-ci.org/makemek/Modcolle)
[![Coverage Status](https://coveralls.io/repos/github/makemek/Modcolle/badge.svg?branch=dev)](https://coveralls.io/github/makemek/Modcolle?branch=dev)

Modcolle is a proxy server for playing [Kantai Collection (艦隊これくしょん ～艦これ～)](http://www.dmm.com/netgame_s/kancolle/).
The main goal of Modcolle is to provide in-game translation displaying in the flash client directly.
Unlike [Kancolle Viewer](https://github.com/Grabacr07/KanColleViewer) that wraps IE browser and listens to outgoing traffic, Modcolle, acting as a web server, makes Kancolle send all HTTP requests to Modcolle instead.
Thus, it has the opportunity to process requests before forwarding them to destinated Kancolle servers or response back to clients.
In this case, Modcolle is a middleman between client and Kancolle servers.
Client can access modcolle without having to configure proxy connection.
No more cookie injection, ~~Japan timezone synchronization~~, or VPN connection in order to play the game.
Modcolle do all those things behind the scenes.
Since the processes involves calling DMM API directly, no bloating javascript is executed, it is faster than accessing from [DMM website](http://www.dmm.com).
Modcolle also distributes game asset requests .png, .mp3, and .swf among multiple Kancolle servers to relief traffic bottle neck.
Ensures that all assets will be delivered as fast as possible.

## Feature
* Enjoy playing the game on full screen
* ~~Place custom content to be displayed inside a game such as ships, musics, or any game assets similar to [KancolleCacher](https://github.com/df32/KanColleCacher).~~ **this feature is suspeneded, will implement sometime later**
* Display translations inside the game **Just for a proof of concept. No full translation are implemented in this release**
* Realtime logging. See what Kancolle makes requests to the server including decoded API requests and responses for statistical analysis

## Installation
Modcolle is a Node.js server application, not a desktop application.
There is no executable launcher.
It is intended to run on a web server.
However, you can setup your machine as a web server to run Modcolle locally.

Modcolle needs to run together with [Nginx](https://www.nginx.com/) in order to operate normally.
Please follows steps below to run Modcolle

### Docker

1. Install [Docker](https://docs.docker.com/engine/installation/) and [Docker Compose](https://docs.docker.com/compose/install/)
2. Clone this repository `git clone git@github.com:makemek/Modcolle.git` and run `npm install`
3. Open terminal and run `docker-compose up`. You can add `-d` flag to detach the process and run as a daemon.
4. Open a web browser and type `localhost` in the url

### Manual Installation

1. Install Nginx and Node.js (v. 6.9.1)
2. Clone this repository `git clone git@github.com:makemek/Modcolle.git` and run `npm install`
3. Copy `/deployment/nginx/nginx.conf` to `\path\to\nginx\conf`. Make sure to backup nginx.conf first before overwrite.
4. Inside nginx.conf replace `app1` and `DOMAIN_NAME` with `localhost`
5. Start nginx
6. Type `npm install pm2 -g` and run `npm start`
7. Open a browser and type `localhost` in the url

## Starting the Application
### Cluster Mode
This command runs [pm2](https://npmjs.org/packages/pm2) using configuration defined in `env` of [process.yml](https://github.com/makemek/Modcolle/blob/master/process.yml).
```
npm start
```
You can also start the application in production mode by passing `--env production` to pm2.
```
npm start -- --env production
```

### Development Mode
With [nodemon](https://npmjs.org/packages/nodemon) and [browser-sync](https://npmjs.org/packages/browser-sync)
> **In this mode, no environment variables are loaded**.
You can still set them inside a command line if needed.
```
npm run dev

-----------
(optional) configure browser-sync using environment variables before running.
PORT=<port number> # browser-sync app's proxy port; default is 5000
PORT_DEV=<port number> # browser-sync listeing port; default is 3000
```

## Configuration
`process.yml` is PM2's [process file](http://pm2.keymetrics.io/docs/usage/application-declaration/) for Modcolle.
Inside `process.yml`, `env` contains common environment variables required by the application.
`env_production` used in production contains envrionment variables that overrides variables in `env`.

### Environment Variables
- LOGGER_LEVEL: select a [logging level](https://github.com/pinojs/pino/blob/master/docs/API.md#level)
- LOGGER_PRETTY: pretty print log (should be set to `false` in production)
- PORT: application port
- PORT_DEV: **FOR DEVELOPMENT ONLY** will take effect when run `npm run dev`
- KANCOLLE_SERVER_MASTER: A host name for main Kancolle server that have interface`/kcsapi/api_world/get_id` for requesting players world id
- KANCOLLE_SERVER_#: A host name for other Kancolle servers where `#` is `world id`


CHANGELOG.md
# Modcolle Changelog
## 2017.05.01, Version 0.3.2
### Notable Changes
* All environment varaibles are loaded from process.yml which requires pm2 to start
* `cklg` cookie are not required to access DMM game page and will not be not be injected to a request
* No logging router requests.
* Use pino logger over winston for better performance

### Commits
* [[`c9dc676`](https://github.com/makemek/modcolle/commit/c9dc676)] doc: update outdated configuration section
* [[`00881de`](https://github.com/makemek/modcolle/commit/00881de)] doc: start the application in production mode
* [[`aa44463`](https://github.com/makemek/modcolle/commit/aa44463)] fix(package): update pino to version 4.5.0
* [[`5562000`](https://github.com/makemek/modcolle/commit/5562000)] chore(package): update async to version 2.4.0
* [[`1b35907`](https://github.com/makemek/modcolle/commit/1b35907)] config: use default pm2 log dir
* [[`cb52dd4`](https://github.com/makemek/modcolle/commit/cb52dd4)] config: remove LOGGER_ENABLE option
* [[`c56d994`](https://github.com/makemek/modcolle/commit/c56d994)] config: fix can't config LOGGER_ENABLE and LOGGER_PRETTY
* [[`42d7d69`](https://github.com/makemek/modcolle/commit/42d7d69)] src: no need to log http request
* [[`f046d7b`](https://github.com/makemek/modcolle/commit/f046d7b)] src: replace winston logger with pino logger
* [[`da91530`](https://github.com/makemek/modcolle/commit/da91530)] chore(package): update nyc to version 10.3.0
* [[`ca72694`](https://github.com/makemek/modcolle/commit/ca72694)] src: refactor code to be es6 complient
* [[`aa37ff8`](https://github.com/makemek/modcolle/commit/aa37ff8)] build: reduce docker image size
* [[`bad5e15`](https://github.com/makemek/modcolle/commit/bad5e15)] chore(package): update coveralls to version 2.13.1
* [[`b4bba33`](https://github.com/makemek/modcolle/commit/b4bba33)] build: start process.yml using production environment
* [[`1d60216`](https://github.com/makemek/modcolle/commit/1d60216)] build: fix pm2-docker not found config file
* [[`89fbdda`](https://github.com/makemek/modcolle/commit/89fbdda)] src: js class implementation in cookie injector
* [[`087d1d7`](https://github.com/makemek/modcolle/commit/087d1d7)] src: remove injecting 'cklg' cookie
* [[`69222dd`](https://github.com/makemek/modcolle/commit/69222dd)] build: 'gulp build' and 'gulp import' as default task
* [[`bc1586e`](https://github.com/makemek/modcolle/commit/bc1586e)] src: replace sprintf-js with es6 string template
* [[`e9a7153`](https://github.com/makemek/modcolle/commit/e9a7153)] dep: dev package should be in devDependencies
* [[`875fe86`](https://github.com/makemek/modcolle/commit/875fe86)] npm,doc: 'npm start' will launch app using pm2
* [[`f7cce20`](https://github.com/makemek/modcolle/commit/f7cce20)] chore(package): update mocha to version 3.3.0
* [[`aa94733`](https://github.com/makemek/modcolle/commit/aa94733)] doc: remarks when running in development mode
* [[`eac0f9a`](https://github.com/makemek/modcolle/commit/eac0f9a)] pm2,bin,test: load all app's envs from process.yml
* [[`34549ba`](https://github.com/makemek/modcolle/commit/34549ba)] pm2,doc: prefer starting the app with pm2 yml config

## 2017.04.26, Version 0.3.1
### Notable Changes
* Maintenance Countdown Timer ([#16](https://github.com/makemek/Modcolle/pull/16))
appears when `Maintenance.IsDoing` in [Kancolle's config](http://203.104.209.7/gadget/js/kcs_const.js) set to `1` and `Maintenance.EndDateTime` is greater than client's machine time.
* `kancolle` folder removed ([#24](https://github.com/makemek/Modcolle/pull/24))
Modding content by redirecting url has been suspended.
If implement, they will be moved to a dedicated microservice.
* Use Node Alpine docker image ([#25](https://github.com/makemek/Modcolle/pull/25))
Smaller image size

### Commits
* [[`d8e1fda`](https://github.com/makemek/Modcolle/commit/d8e1fda)] update readme.md
* [[`a8ef58e`](https://github.com/makemek/Modcolle/commit/a8ef58e)] use node alpine image ([#25](https://github.com/makemek/Modcolle/pull/25))
* [[`95f97c3`](https://github.com/makemek/Modcolle/commit/95f97c3)] remove unused Kancolle folder ([#24](https://github.com/makemek/Modcolle/pull/24))
* [[`94b8340`](https://github.com/makemek/Modcolle/commit/94b8340)] fix(package): update url-join to version 2.0.1 ([#23](https://github.com/makemek/Modcolle/pull/23))
* [[`49f9f47`](https://github.com/makemek/Modcolle/commit/49f9f47)] chore(package): update nock to version 9.0.13 ([#22](https://github.com/makemek/Modcolle/pull/22))
* [[`892588e`](https://github.com/makemek/Modcolle/commit/892588e)] chore(package): update babel-preset-env to version 1.4.0 ([#21](https://github.com/makemek/Modcolle/pull/21))
* [[`c132000`](https://github.com/makemek/Modcolle/commit/c132000)] fix(package): update url-join to version 2.0.0 ([#20](https://github.com/makemek/Modcolle/pull/20))
* [[`db982b0`](https://github.com/makemek/Modcolle/commit/db982b0)] eslint: fix single quote error
* [[`e4feb81`](https://github.com/makemek/Modcolle/commit/e4feb81)] Maintenence countdown timer ([#16](https://github.com/makemek/Modcolle/pull/16))

## 2017.04.23, Version 0.3.0
### Notable Changes
* Frontend
  - Fallback support for font-awesome. Load font-awesome from their CDN. If fails, load them from the server
  - CSS and JS are now minified and uglified on build
  - Bigger launcher button
  - In development mode, sync css and js automaticaaly on file changes to browser
* Can be launched using PM2
* Incorporate new sinon v2 API changes
* `index.html` rendered statically

### Commits
* [[`02ded80`](https://github.com/makemek/Modcolle/commit/02ded80)] Update test code to new sinon v2 API (#18)
* [[`91087c4`](https://github.com/makemek/Modcolle/commit/91087c4)] Update dependencies to enable Greenkeeper <F0><9F><8C><B4> (#17)
* [[`648aba2`](https://github.com/makemek/Modcolle/commit/648aba2)] reuse eslint config from modcolle-eslint-config (#15)

gulpfile.js
'use strict'

const appPort = process.env.PORT || '5000'
const devPort = process.env.PORT_DEV || '3000'
const css = 'src/views/*.css'
const js = 'src/views/js/*.js'
const html = 'src/views/**/*.html'
const htmlTemplate = 'src/views/**/*.hbs'
const publicAssets = 'src/views/public'
const destination = {
  css: publicAssets + '/css',
  fonts: publicAssets + '/fonts',
  js: publicAssets + '/js'
}

const gulp = require('gulp')
const postcss = require('gulp-postcss')
const autoprefixer = require('autoprefixer')
const cleanCSS = require('gulp-clean-css')
const rename = require('gulp-rename')
const bs = require('browser-sync').create()
const nodemon = require('gulp-nodemon')
const uglify = require('gulp-uglify')
const babel = require('gulp-babel')
const gutil = require('gulp-util')
const source = require('vinyl-source-stream')
const browserify = require('browserify')
const buffer = require('vinyl-buffer')

gulp.task('default', ['build', 'import'])

gulp.task('build', ['build:css', 'build:js'])

gulp.task('build:css', () => {
  return gulp.src(css)
  .pipe(rename({suffix: '.min'}))
  .pipe(postcss([autoprefixer()]))
  .pipe(cleanCSS())
  .pipe(gulp.dest(destination.css))
  .pipe(bs.stream())
})

gulp.task('build:js', () => {
  browserify('src/views/js/')
  .bundle()
  .on('error', error => gutil.log(error))
  .pipe(source('bundle.min.js'))
  .pipe(buffer())
  .pipe(babel())
  .pipe(uglify())
  .pipe(gulp.dest(destination.js))
  .pipe(bs.stream())
})

gulp.task('import', ['font-awesome'])

gulp.task('font-awesome', () => {
  gulp.src('node_modules/font-awesome/css/font-awesome.min.css')
  .pipe(gulp.dest(destination.css))
  gulp.src('node_modules/font-awesome/fonts/*')
  .pipe(gulp.dest(destination.fonts))
})

gulp.task('nodemon', done => {
  let start = false
  return nodemon({
    script: 'bin/www',
    ignore: ['gulpfile.js', 'node_modules/', 'src/views/'],
    env: {PORT: appPort}
  })
  .on('start', () => {
    if(!start) {
      start = true
      done()
    }
  })
})

gulp.task('browser-sync', ['nodemon'], () => {
  bs.init({
    proxy: `localhost:${appPort}`,
    port: devPort,
    open: false
  })

  gulp.watch(css, ['build:css'])
  gulp.watch(js, ['build:js'])
  gulp.watch(html, bs.reload)
  gulp.watch(htmlTemplate, bs.reload)
})
